### ISSUE TITLE
[@typescript-eslint/no-unnecessary-condition] The rule reports false positives object field is optional in typescript

### ISSUE THREAD

**Issue by asheeshims** [typescript-eslint/typescript-eslint#1594]
[@typescript-eslint/no-unnecessary-condition] The rule reports false positives object field is optional in typescript

**Repro**

```JSON
{
  "rules": {
            "no-useless-constructor": "off",
        "import/no-unresolved": "off",
        "no-constant-condition": "off",
        "no-unused-expressions": "off",
        "no-unnecessary-type-assertion": "off",
        "@typescript-eslint/no-non-null-assertion": "error",
        "@typescript-eslint/prefer-nullish-coalescing": "off",
        "@typescript-eslint/prefer-optional-chain": "off",

        "@typescript-eslint/no-extra-non-null-assertion": "error",
        "@typescript-eslint/no-non-null-asserted-optional-chain": "error",
        "@typescript-eslint/no-unnecessary-condition": ["error", { "ignoreRhs": true }],
        "@typescript-eslint/no-unused-expressions": ["error", { "allowShortCircuit": true, "allowTernary": true, "allowTaggedTemplates": true }],
        "@typescript-eslint/no-unnecessary-type-arguments": "error",
        "@typescript-eslint/no-unnecessary-type-assertion": "error",
  }
}
```

```TS
   interface Emp {
        fname: string;
        lname?: string;
    }
    
    const emp = { fname: 'John', lname: 'Doe'};
    const lnameLen = emp.lname?.length; 
```

**Expected Result**

**Actual Result**
  The plugin shows error at **emp.lname?.length** at **lname?.length** reporting that ESLint: Unnecessary optional chain on a non-nullish value(@typescript-eslint/no-unnecessary-condition) while **lname** is optional. The same behavior happen even when i do not set **lname** in the **emp** object.

**Additional Info**
```.eslintrc.js
module.exports = {
    root: true,
    parser: "@typescript-eslint/parser",
    plugins: ["import", "react", "jsx-a11y", "jest", "@typescript-eslint"],
    parserOptions: {
        sourceType: "module",
        ecmaFeatures: {
            jsx: true,
            generators: false,
            objectLiteralDuplicateProperties: false
        },
        ecmaVersion: 2018,
        project: "tsconfig.json",
        tsconfigRootDir: __dirname,
    },
    //extends: ["plugin:@typescript-eslint/eslint-recommended", "plugin:@typescript-eslint/recommended"],
    rules: {
        "no-useless-constructor": "off",
        "import/no-unresolved": "off",
        "no-constant-condition": "off",
        "no-unused-expressions": "off",
        "no-unnecessary-type-assertion": "off",
        "@typescript-eslint/no-non-null-assertion": "error",
        "@typescript-eslint/prefer-nullish-coalescing": "off",
        "@typescript-eslint/prefer-optional-chain": "off",

        "@typescript-eslint/no-extra-non-null-assertion": "error",
        "@typescript-eslint/no-non-null-asserted-optional-chain": "error",
        "@typescript-eslint/no-unnecessary-condition": ["error", { "ignoreRhs": true }],
        "@typescript-eslint/no-unused-expressions": ["error", { "allowShortCircuit": true, "allowTernary": true, "allowTaggedTemplates": true }],
        "@typescript-eslint/no-unnecessary-type-arguments": "error",
        "@typescript-eslint/no-unnecessary-type-assertion": "error",
    },
    env: {
        es6: true,
        node: true,
        jest: true,
        jasmine: true,
        browser: true,
    },
    settings: {
        'import/parsers': {
            '@typescript-eslint/parser': ['.ts', '.tsx'],
        },
        "import/resolver": {
            node: {
                extensions: [".ts", ".tsx"]
            }
        },
        react: {
            pragma: "React",
            version: "detect"
        },
        propWrapperFunctions: ["forbidExtraProps", "exact", "Object.freeze"],
        "import/extensions": [".js", ".mjs", ".jsx", ".ts", ".tsx"],
        "import/core-modules": [],
        "import/ignore": ["node_modules", "\\.(coffee|scss|css|less|hbs|svg|json)$"]
    },
};
```

```tsconfig.json
{
  "compilerOptions": {
    "allowJs": true,
    "outDir": "./dist/",
    "incremental": true,
    "noImplicitAny": true,
    "sourceMap": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": false,
    "forceConsistentCasingInFileNames": true,
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "target": "es5",
    "jsx": "react",
    "module": "esnext",
    "moduleResolution": "node",
    "typeRoots": [
      "./node_modules/@types",
      "./src/common/declarations"
    ],
    "lib": [
      "esnext.asynciterable",
      "dom",
      "es5"
    ],
    "types": [
      "date-fns",
      "es6-promise",
      "facebook-js-sdk",
      "googlemaps",
      "jsdom",
      "lodash",
      "lodash-webpack-plugin",
      "react",
      "react-addons-create-fragment",
      "react-async-script-loader",
      "react-transition-group",
      "react-dates",
      "react-dom",
      "react-modal",
      "react-slick",
      "gapi",
      "gapi.auth2",
      "google-map-react",
      "google-maps-promise",
      "object-fit-images",
      "rc-tooltip",
      "rc-slider",
      "react-select3",
      "react-copy-to-clipboard",
      "react-lazy-load",
      "react-markdown",
      "react-swipe",
      "react-helmet",
      "react-visibility-sensor",
      "react-youtube",
      "matchmedia-polyfill"
  ]
  },
  "awesomeTypescriptLoaderOptions": {
    "useWebpackText": true,
    "useTranspileModule": true,
    "doTypeCheck": true,
    "forkChecker": true
  },
  "include": [
    "src/**/*.ts",
    "src/**/*.tsx",
  ],
  "exclude": [
    "/node_modules/",
    "**/*.test.ts",
    "**/*.test.tsx",
    "**/*Mock.ts",
    "**/*mockData.ts",
    "**/*Mock.tsx",
    "**/*mockData.tsx",
    "src/**/mocks/"
  ]
}

```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.19.2"` |
| `@typescript-eslint/parser`        | `2.19.2` |
| `eslint-config-airbnb`                 | `18.0.1` |
| `TypeScript`                       | `X.Y.Z` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `10.15.3` |
| `npm`                              | `6.4.1` |
| `typescript`                      | `3.7.3` |

**Comment by bradzacher** [comment]
This is completely correct.
You are not actually attaching your type to your variable.

```ts
   interface Emp {
        fname: string;
        lname?: string;
    }
    
    const emp = { fname: 'John', lname: 'Doe'};
//        ^^^^^ no explicit type, so it isn't of type Emp, it's of type { fname: string; lname: string }
    const lnameLen = emp.lname?.length; 
```

If you explicitly type `emp` as `Emp`, then the rule no longer reports on the object, as it is correctly typed as optional

**Comment by asheeshims** [comment]
Sorry for the typo but this code also does not work eirther:

`const emp: Emp = { fname: 'John', lname: 'Doe' };
 const lnameLen = emp.lname?.length;`

**Comment by asheeshims** [comment]
> This is completely correct.
> You are not actually attaching your type to your variable.
> 
> ```ts
>    interface Emp {
>         fname: string;
>         lname?: string;
>     }
>     
>     const emp = { fname: 'John', lname: 'Doe'};
> //        ^^^^^ no explicit type, so it isn't of type Emp, it's of type { fname: string; lname: string }
>     const lnameLen = emp.lname?.length; 
> ```
> 
> If you explicitly type `emp` as `Emp`, then the rule no longer reports on the object, as it is correctly typed as optional

Sorry for the typo but this code also does not work either:
```
const emp: Emp = { fname: 'John', lname: 'Doe' }; 
const lnameLen = emp.lname?.length;
```
Can you help me with a working example. How should i create the employee object to make this rule work.
I pulled the code for '@typescript-eslint' and the same code works as expected in your test cases.

**Comment by bradzacher** [comment]
It should (and does in my tests), but it doesn't because you don't use `strictNullChecks`.

With `strictNullChecks` off, typescript essentially erases all `null` and `undefined` types from your code.

**Comment by asheeshims** [comment]
Thanks a ton!!
It work with **strictNullChecks** option.

```
const emp: Emp = { fname: 'John', lname: 'Doe' };
const fnameLen = emp?.fname.length;
```
Now it should fail at **emp?** because emp object is not null, but it does not.
What is wrong in my config for this.

**Comment by bradzacher** [comment]
![image](https://user-images.githubusercontent.com/7462525/74390636-bdaf7280-4db6-11ea-90a3-90b66be5db9f.png)

Works fine in the tests.
If you're playing with this snippet in our unit tests, note that you need the interface declaration. Without the interface, TS resolves the type of `Emp` to `any`, so the rule treats the `?.` as valid.

**Comment by asheeshims** [comment]
Thanks it started to work without any change.
The rule is working as expected.
